### PR TITLE
Add compound composition and decomposition

### PR DESCRIPTION
**Pull request by brownbat**
Adds compound composition and decomposition to address the issue I filed in #36 (admittedly an issue you may or may not even want to resolve!)

Tried to follow your style and only make a few changes. You might prefer a different model for the constants though. Also catch my NOTE: at 326, I'm stretching your hcj_to_jamo model, which would only be an issue if you make that jamo "position" requirement more strict.

Take or leave, and thanks for the project!

### CODE REVIEW

**Comment by brownbat** [comment]
Made some changes based on the feedback, I think they folded in up above.

I'm new to git, so I'm not 100% sure if I'm updating the PR correctly... For one, I don't know why I ended up with "patch-4" where all these changes live, instead of just in "master" on my fork, but fun times, learning as I go.

Another confession, I've never actually written tests before. I'm learning a ton from google and your code actually, should be able to take a stab here soon, just bear with me. :)

Thanks for considering the changes.

**Comment by brownbat** [comment]
Hmm, my tests aren't running. I think it's not importing the new jamo.py to find the new functions, not sure. Will have to dig around on this a bit...

Also - you don't have to go with the unicodedata import, I can rewrite to follow the .json you implemented, but thought you might find it nice to see unicode names built-in.

File: jamo/jamo.py [hunk 1]
```diff
@@ -146,6 +181,34 @@ def is_hangul_char(character):
...
 
+def is_jamo_compound(character):
+    """Test if a single character is a compound, i.e., a consonant 
+    cluster, double consonant, or dipthong.
+    """
+    if len(character) != 1:
+        raise TypeError('is_jamo_compound() expected a single character')
+    if is_jamo(character):
+        character_name = unicodedata.name(character)
+        if "-" in character_name:
```
Comment (JDongian) [inline jamo/jamo.py]: All this seems a bit hacky (esp. the `W`). I would have just made a set of all the compounds and done `if chr in COMPOUNDS` or something like that. However, if it passes the tests I'm okay with this for now.

File: jamo/jamo.py [hunk 2]
```diff
@@ -287,10 +348,45 @@ def j2h(lead, vowel, tail=0):
...
 
 
+def decompose_jamo(compound):
+    """Return a tuple of jamo character constituents of a compound.
+    Note: Non-compound characters are echoed back.
+
+    WARNING: Archaic jamo compounds will raise NotImplementedError.
+    """
+    if len(compound) != 1:
+      raise TypeError("decompose_jamo() expects a single character,",
```
Comment (JDongian) [inline jamo/jamo.py]: You can run `make lint` to check for inconsistencies like these. `flake8` will need to be installed.

File: tests/test_jamo.py [hunk 3]
```diff
@@ -527,6 +528,184 @@ def test_j2h(self):
...
+        test_chars = ["ㄸ","ㅢ"]
+        target_chars = [("ㄷ","ㄷ"), ("ㅡ","ㅣ")]
+        
+        test_chars_idempotent = itertools.chain(invalid_hangul, 
+                                invalid_other)
+        target_chars_idempotent = test_chars_idempotent
+        
+        #invalid
+        invalid_strings = ["ab","ㄸㄲ"]
+
```
Comment (JDongian) [inline tests/test_jamo.py]: avoid double newlines within a method

**Comment by JDongian** [comment]
Ah, you didn't set up the imports in __init__.py. I can do that real quick.

File: jamo/jamo.py (hunk 2 above)
Comment (brownbat) [reply jamo/jamo.py]: Wow, flake8 is so helpful. Rereading PEP 8 I realized how off I was. Even, my long line marker was way up at 72 for some reason...

I hacked tests to work locally yesterday by uninstalling jamo and forcing the import to look at a relative directory. I will now sheepishly read about how init files work and how to actually do this the right way though!

I'll have some updates soon, style and bugfixes the tests revealed.

**Comment by brownbat** [comment]
1. The __init__.py changes alone did not let me run tests locally. I temporarily used a relative import as a temporary hack, then put the import back before submitting, but I think I'm still missing something about how to get the tests to find the right jamo.py.

2. Still a couple things on my todo list. I'd like to go back and simplify my constants. Since you have a function to determine if something's modern or not,  you shouldn't really need separate _MODERN_ constants, just a constant with all the compounds. A full compounds constant will also help me write a full compounds dictionary, which in turn will let me replace the admittedly "hacky" function.

3. When I tried to bring in your __init__.py changes, I ended up in vimdiff and panicked a bit, [as people do](https://stackoverflow.blog/2017/05/23/stack-overflow-helping-one-million-developers-exit-vim/), so backed out and just implemented them fresh by hand. In other fun missteps, I now have both a patch-4 branch and a patch_4 branch.

So, also on my todo list is: dig around for a deeper tutorial on git, or just play around with it more so I actually know what I'm doing.

What do you use as a tool for merges, to reconcile differences? Github's web interface is actually really intuitive, so I'm falling back on that, but I feel a strange guilt not running something locally.

**Comment by JDongian** [comment]
1. really...? they would not even _run_..? It can be tricky with the imports; remember you can always use the full path `jamo.jamo.<name>`.
2. sounds fine
3. `:wq` writes, then quits. `:q` quits. Set your default EDITOR to something else if you don't git to use vim.

Glad you're learning how to work with python and use git! I do everything on the command line, and handle merge conflicts manually (in `vim`). It's not hard -- just search for the `======` and fix each one. Github makes it easy via web, and it works just fine, too.

**Comment by brownbat** [comment]
Yeah, to be specific, it couldn't load the module. test_jamo.py just fails immediately with:
```
Traceback (most recent call last):
  File "test_jamo.py", line 5, in <module>
    import jamo
ModuleNotFoundError: No module named 'jamo'
```

A relative import works fine, ie:
```
import os
import sys
# set current working directory to location of test_jamo.py to make relative directory ("..")
#   consistent regardless of where the file is launched from
original_cwd = os.getcwd()
os.chdir(sys.path[0])
sys.path.append(os.path.join("..","jamo"))
os.chdir(original_cwd)
import jamo
```

```
C:\[...]\python-jamo\tests>python test_jamo.py
..................
----------------------------------------------------------------------
Ran 18 tests in 0.017s

OK
```

Is `os.getcwd()` not putting you inside the tests directory to start?

**Comment by JDongian** [comment]
That shouldn't be happening. From a fresh clone:
```
$ git status
On branch patch-4
Your branch is up-to-date with 'brownbat/patch-4'.

nothing to commit, working tree clean
$ ls
LICENSE		Makefile	README.rst	docs		docs-kr		jamo		setup.py	tests		tools
$ make test
nosetests --with-coverage --cover-package=jamo
EEE............E..
======================================================================
ERROR: compose_jamo tests
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/tmp/python-jamo/tests/test_jamo.py", line 630, in test_compose_jamo
    _stderr = jamo.stderr
AttributeError: module 'jamo' has no attribute 'stderr'

======================================================================
ERROR: decompose_jamo tests
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/tmp/python-jamo/tests/test_jamo.py", line 579, in test_decompose_jamo
    _stderr = jamo.stderr
AttributeError: module 'jamo' has no attribute 'stderr'

======================================================================
ERROR: get_jamo_class tests
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/tmp/python-jamo/tests/test_jamo.py", line 227, in test_get_jamo_class
    _stderr = jamo.stderr
AttributeError: module 'jamo' has no attribute 'stderr'

======================================================================
ERROR: jamo_to_hangul tests
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/tmp/python-jamo/tests/test_jamo.py", line 502, in test_jamo_to_hangul
    _stderr = jamo.stderr
AttributeError: module 'jamo' has no attribute 'stderr'

Name           Stmts   Miss  Cover
----------------------------------
jamo.py            2      0   100%
jamo/jamo.py     172     15    91%
----------------------------------
TOTAL            174     15    91%
----------------------------------------------------------------------
Ran 18 tests in 0.063s

FAILED (errors=4)
make: *** [test] Error 1
```

**Comment by JDongian** [comment]
Well, I did run `git checkout brownbat patch-4`, after setting my `brownbat` remote to point to your fork. As long as you are on branch `patch-4` you should get the same output.

**Comment by JDongian** [comment]
Thanks, this is great! I removed the path fix, since it's not required for the CI.

**Pull request merged**